sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+        # For each element (resource) in the resources sequence, we extract the resource id
+        resource_list = map(lambda resource: resource['id'], resources)
+        return list(resource_list)
+```
+
+See the  [Client class](https://github.com/cyberark/conjur-api-python3/blob/main/conjur/api/client.py) and [API class](https://github.com/cyberark/conjur-api-python3/blob/main/conjur/api/api.py#L128) and for more details.
+
+1. Multiple responsibilities and not easily expandable.
+
+The *API* class handles multiple responsibilities. The API class handles the building of the request based on the types of information provided by the user *and* the invoking of the call *and* formatting the response.
+
+2. Areas of volatility are not encapsulated.
+
+We can add commands, change the way we present a certain command's output, and/or the types of parameters can manipulate the overall flow. Each command has its own associated function in `client.py` and `api.py` so each time we want to add a new command, we need to re-open multiple classes and add the desired functionality. This overcrowds the *Api*, allowing command-specific logic to spill into a very general class. 
+
+##### Proposed flow
+
+For these reasons, I propose we extract the command-specific logic into its own class. That way, the the *Api* class will be exposed to the specific details of a command, how to package the request, how to make the call, what format the request should be in. This works to isolate the areas of volatility should we require to support different use cases for a command. 
+
+```python
+class ConjurClient()
+	def __init__(...):
+    self.api_client = ApiClient()
+    
+	def list(self, list_constraints: dict = None) -> dict:
+    list_api = ListApi(list_constraints)
+	  return self.list_api.execute(list_constraints)
+
+# interface common to all commands (list, variable, policy, etc)
+class CommandInterface()
+	def build_request()
+  def execute()
+  def format_request()
+  
+class ListApi(CommandInterface)
+	def __init__(api_client = None):
+    self.api_client = api_client
+
+	def build_request(list_constraints: dict = None)
+		# … validate …
+		# … build params for request
+		# … return built list request
+    
+	def format_request()
+		# … format request (JSON, string, etc)
+    
+  def execute(list_constraints: dict = None)
+		# builds the request
+    self.build_request(list_constraints)  
+    # makes the API call
+		response = self.api_client.call_api(…)
+    # formats response to List/JSON/etc
+		return self.format_request(response)
+  
+class ApiClient():
+  def __init__(...):
+     # …    
+
+  def call_api(...)
+  	# 1. build query and headers
+		# 2. escape data
+    # 3. perform request
+    # 4. return response
+
+```
+
+
+
+### Full proposed SDK flow
+
+![SDKReactorFlow](design/assets/SDKReactorFlow.png)
+
+At a high level,
+
+1. The user will initalize a Configuration and Credential object, *separate* from the Client
```

Comment:
See my comment reg. having a documented fetching logic so we won't need to initialize the configuration object